### PR TITLE
feat: add v4 initial version, with options and restructured package

### LINKED ISSUE #114: fatal error: concurrent map writes
I have a goroutine in a for loop that runs `faker.FakeData`, here a sample code:
```go
package main

import (
	"reflect"
	"github.com/bxcodec/faker"
)

type User struct {
	Username string `faker:"first_name"`
	Email    string `faker:"email,unique"`
}

func main() {
	user := User{}
	modelType := reflect.TypeOf(user)
	for i := 0; i < 1000; i++ {
		go func() {
			for j := 0; j < 1000; j++ {
				modelCopy := reflect.New(modelType).Interface() // copy model
				faker.FakeData(modelCopy)
			}
		}()
	}
}

```

I get the follow error:
```
$ go run bin/faker.go 
fatal error: concurrent map read and map write

goroutine 19 [running]:
runtime.throw(0x1122ca3, 0x21)
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/panic.go:1114 +0x72 fp=0xc000221760 sp=0xc000221730 pc=0x1030a22
runtime.mapaccess1_faststr(0x10f41a0, 0xc00009a600, 0x10e3609, 0x5, 0x10eb480)
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/map_faststr.go:21 +0x43c fp=0xc0002217d0 sp=0xc000221760 pc=0x10120fc
github.com/bxcodec/faker.getValue(0x10fb380, 0xc00020e580, 0x199, 0x10fb301, 0x10fb380, 0xc00020e580, 0xc00020e500)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:422 +0x1e2e fp=0xc000221b30 sp=0xc0002217d0 pc=0x10cbade
github.com/bxcodec/faker.getValue(0x10e58e0, 0xc00020e540, 0xc000221f48, 0x107ab78, 0x10fb380, 0xb7a0, 0x10e58e0)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:356 +0x174a fp=0xc000221e90 sp=0xc000221b30 pc=0x10cb3fa
github.com/bxcodec/faker.FakeData(0x10e58e0, 0xc00020e540, 0x16, 0xc00020e501)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:281 +0x1c6 fp=0xc000221f88 sp=0xc000221e90 pc=0x10c9896
main.main.func1(0x1145580, 0x10fb380)
	/Users/pioz/Code/go/src/github.com/pioz/ct/bin/faker.go:23 +0x81 fp=0xc000221fd0 sp=0xc000221f88 pc=0x10da051
runtime.goexit()
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000221fd8 sp=0xc000221fd0 pc=0x105d6a1
created by main.main
	/Users/pioz/Code/go/src/github.com/pioz/ct/bin/faker.go:20 +0xaf

goroutine 1 [runnable]:
runtime.gopark(0x0, 0x0, 0xc000001008, 0x1)
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/proc.go:287 +0x130
runtime.main()
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/proc.go:222 +0x28c
runtime.goexit()
	/usr/local/Cellar/go/1.14.1/libexec/src/runtime/asm_amd64.s:1373 +0x1

goroutine 20 [runnable]:
github.com/bxcodec/faker/support/slice.ContainsValue(0xc000226000, 0xaf, 0x100, 0x10eb480, 0xc00010d5b0, 0xc00010d5b0)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/support/slice/helpers.go:21 +0x82
github.com/bxcodec/faker.getValue(0x10fb380, 0xc000115240, 0x199, 0x10fb301, 0x10fb380, 0xc000115240, 0xc0001151c0)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:422 +0x1e6b
github.com/bxcodec/faker.getValue(0x10e58e0, 0xc000115200, 0xc000195f48, 0x107ab78, 0x10fb380, 0xb7a0, 0x10e58e0)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:356 +0x174a
github.com/bxcodec/faker.FakeData(0x10e58e0, 0xc000115200, 0x16, 0xc000115201)
	/Users/pioz/Code/go/src/github.com/bxcodec/faker/faker.go:281 +0x1c6
main.main.func1(0x1145580, 0x10fb380)
	/Users/pioz/Code/go/src/github.com/pioz/ct/bin/faker.go:23 +0x81
created by main.main
	/Users/pioz/Code/go/src/github.com/pioz/ct/bin/faker.go:20 +0xaf
...
```
If I remove the `go` keyword all works fine.
If I remove the `unique` option in the `Email` field tag all works fine.

### LINKED ISSUE #144: Max depth parameter
I am trying to fill up the structure that has infinite loop inside.
That would be awesome to set a limiter that stops after a few loops

### LINKED ISSUE #153: `slice_len` only works for `string`
Right now, `slice_len` does not actually work for a slice of integers/bytes, despite what the example has: https://github.com/bxcodec/faker/blob/master/example_with_tags_slicelength_test.go

If you add a slice of integers to the test: https://github.com/bxcodec/faker/blob/master/faker_test.go#L650

It will fail with "Invalid Tags" here: https://github.com/bxcodec/faker/blob/master/faker.go#L921

The reason you don't see similar with Strings is because the similar check in `extractStringFromTag`, we always have `lang` set: https://github.com/bxcodec/faker/blob/master/faker.go#L858

I wasn't sure how best to approach this, whether we should just have a check accepting `slice_len` as an additional tag in `extractNumberFromTag` or some other cleaner way, but dem's de facts.

### PR DESCRIPTION
**Pull request by bxcodec**
fix #153 
fix #144 
fix #114

### CODE REVIEW

File: datetime.go [hunk 1]
```diff
@@ -829,11 +819,11 @@ func (d DateTime) TimePeriod(v reflect.Value) (interface{}, error) {
 }
 
 // Timeperiod get timeperiod randomly in string (AM/PM)
-func Timeperiod() string {
+func Timeperiod(opt *options.Options) string {
```
Comment (ridhoperdana) [inline datetime.go:822]: why do you use pointer?

File: options/options.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	"errors"
+	"fmt"
+	"reflect"
+
+	fakerErrors "github.com/bxcodec/faker/v4/errors"
+	"github.com/bxcodec/faker/v4/interfaces"
+)
+
+// Options represent all available option for faker.
+type Options struct {
```
[the remark this answers is not in the document]
Comment (bxcodec) [reply options/options.go:13]: Right.... 
Let me think how to make this

File: address.go [hunk 3]
```diff
@@ -2,26 +2,16 @@ package faker
...
 
 // GetAddress returns a new Addresser interface of Address
 func GetAddress() Addresser {
-	mu.Lock()
-	defer mu.Unlock()
-
-	if address == nil {
-		address = &Address{}
-	}
+	address := &Address{}
```
Comment (ridhoperdana) [inline address.go:11]: if we remove the mutex usage, is this still safe for concurrent usage?

Comment (wolf-joe) [reply address.go:11]: Yes, `address` now become a local variable. Every goroutine will get an independent instance.

File: options/options.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+// WithRecursionMaxDepth used for configuring the max depth of nested struct for faker
+func WithRecursionMaxDepth(depth uint) OptionFunc {
+	return func(oo *Options) {
+		if oo.MaxDepthOption == nil {
+			oo.MaxDepthOption = &MaxDepthOption{
+				recursionMaxDepth: 1, // default
+				typeSeen:          make(map[reflect.Type]int, 1),
+			}
+		}
+		if depth > 0 {
```
Comment (wolf-joe) [inline options/options.go:124]: I think `depth` can set 0, which means disabling recursion.

Comment (bxcodec) [reply options/options.go:124]: Alright, so the default valuer is `0` is that correct?
I will address all the comment tomorrow weekend

File: faker_test.go [hunk 5]
```diff
@@ -2319,13 +2272,13 @@ func TestFakeData_RecursiveType(t *testing.T) {
 		return string(data)
 	}
 	node1 := BinaryTreeNode{}
-	if err := FakeData(&node1, WithRecursionMaxDepth(2)); err != nil {
+	if err := FakeData(&node1, options.WithRecursionMaxDepth(2)); err != nil {
 		t.Errorf("%+v", err)
 		t.FailNow()
```
[the remark this answers is not in the document]
Comment (wolf-joe) [reply faker_test.go:2280]: already fix

File: options/options.go (hunk 4 above)
Comment (wolf-joe) [reply options/options.go]: Not really, allow recursive one times is more suitable for my use case

**Comment by bxcodec** [comment]
@wolf-joe  can you help to re-review again? 

I decided to keep Options publicly available, the reason is just it's the faker option, and people can read it as they want.
So they can be aware what are the options that we provided.

File: example_single_fake_data_test.go [hunk 6]
```diff
@@ -1,71 +1,70 @@
...
+	"github.com/bxcodec/faker/v4/options"
+)
 
 // Single fake function can be used for retrieving particular values.
 func Example_singleFakeData() {
 
 	// Address
-	faker.Latitude()  // => 81.12195
-	faker.Longitude() // => -84.38158
+	faker.Latitude(options.DefaultOption())  // => 81.12195
```
[the remark this answers is not in the document]
Comment (bxcodec) [reply example_single_fake_data_test.go]: PR updated. please help to re-review again

File: faker.go [hunk 7]
```diff
@@ -318,86 +215,40 @@ func ResetUnique() {
...
-}
-
-// SetRandomMapAndSliceSize sets the size for maps and slices for random generation.
-// deprecates, currently left for old version usage
-func SetRandomMapAndSliceSize(size int) error {
-	return SetRandomMapAndSliceMaxSize(size)
-}
-
-// SetRandomMapAndSliceMaxSize sets the max size for maps and slices for random generation.
-func SetRandomMapAndSliceMaxSize(size int) error {
```
Comment (wolf-joe) [inline faker.go:357]: I prefer to keep those set function, it provides global setting and back forward compatibility

Comment (bxcodec) [reply faker.go:357]: I don't this is possible, since we remove the global variables, and using options

Comment (bxcodec) [reply faker.go:357]: And since we're introducing major version, breaking changes is acceptable. 

And no need to be 100% backward compatibility

**Comment by bxcodec** [comment]
I'm merging this now and will release a v4.0.0-beta version. 
And will do fixing for any incoming issues after this breaking releases.

**Pull request merged**